pulp-glue/pulp_glue/common/authentication.py

```diff
@@ -0,0 +1,94 @@
+import json
+import typing as t
+from datetime import datetime, timedelta
+from pathlib import Path
+
+import click
+import requests
+
+
+class OAuth2ClientCredentialsAuth(requests.auth.AuthBase):
+    def __init__(
+        self,
+        client_id: str = "",
+        client_secret: str = "",
+        token_url: str = "",
+        scopes: t.List[str] = [],
+        *args: t.List[t.Any],
+        **kwargs: t.Dict[t.Any, t.Any],
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+        self.scopes = scopes
+        self.token: t.Dict[t.Any, t.Any] = {}
+
+    def __call__(self, request: requests.PreparedRequest) -> requests.PreparedRequest:
+        self.retrieve_local_token()
+
+        access_token = self.token.get("access_token")
+        request.headers["Authorization"] = f"Bearer {access_token}"
+
+        request.register_hook("response", self.handle401)  # type: ignore
+
+        return request
+
+    def handle401(
+        self, response: requests.Response, **kwargs: t.Dict[t.Any, t.Any]
+    ) -> requests.Response:
+        if response.status_code != 401:
+            return response
+
+        self.retrieve_local_token()
+        if self.is_token_expired():
+            self.retrieve_token()
+
+        response.content
+        prep = response.request.copy()
+
+        access_token = self.token.get("access_token")
+        prep.headers["Authorization"] = f"Bearer {access_token}"
+
+        _response: requests.Response = response.connection.send(prep, **kwargs)  # type: ignore
+        _response.history.append(response)
+        _response.request = prep
+
+        return _response
+
+    def is_token_expired(self) -> bool:
+        if self.token:
+            issued_at = datetime.fromisoformat(self.token["issued_at"])
+            expires_in = timedelta(seconds=self.token["expires_in"])
+            token_timedelta = issued_at + expires_in - timedelta(seconds=5)
+
+            if token_timedelta >= datetime.now():
+                return False
+
+        return True
```

Comment:
Instead of this rather long check, let's persist `expires_at` on the object and replace the call to this function with `if expires_at is None or datetime.now() > expires_at`